File: Codes/HappyNumber/HappyNumber.hs
import Data.Char (digitToInt)
import Control.Monad (join)

is_happy_number n history
    | sumnum == 1 = True
    | sumnum `elem` history = False
    | otherwise = is_happy_number sumnum (sumnum:history)
    where sumnum = sum [(digitToInt x)^2 | x<-show(n)]

test_happy_number n = join [show(n), " is ", (if is_happy_number n [] then "a Happy" else "an Unhappy"), " number."]

main = do
    mapM_ putStrLn $ map test_happy_number [3,7,4,13]

File: Codes/HappyNumber/HappyNumber.py
# find happy number
def is_happy_num(s0, history=[]):
    s1 = sum(map(lambda x:x*x, [int(i) for i in repr(s0)]))    
    if s1 == 1:
        return True
    elif s1 in history:
        return False
    else:
        history.append(s1)
        return is_happy_num(s1)

# Test case
for i, num in enumerate([3,7,4,13]):
    case = "Case #%d: %d is " % (i, num)
    result = "Happy number." if is_happy_num(num, []) else "an Unhappy number."
    print(case + result)